rust/core-lib/src/plugins/manager.rs

```diff
@@ -116,93 +135,129 @@ impl <W: Write + Send + 'static>PluginManager<W> {
-        //TODO: error handling: this should maybe have a completion callback with a Result
-        let plugin_id = self.next_plugin_id();
-        let buffer_id = self.buffers.buffer_for_view(view_id);
-        if let Some(buffer_id) = buffer_id {
-            let key = (buffer_id.to_owned(), plugin_name.to_owned());
-            if self.running.contains_key(&key) {
-                return print_err!("plugin {} already running for buffer {}", plugin_name, key.0);
-            }
+    fn start_plugin(&mut self, self_ref: &PluginManagerRef<W>,
+                    view_id: &ViewIdentifier, plugin_name: &str,
+                    init_info: PluginBufferInfo) -> Result<(), Error> {
 
-            let plugin = self.catalog.iter().find(|desc| desc.name == plugin_name)
-                .expect(&format!("no plugin found with name {}", plugin_name));
-
-            let me = self_ref.clone();
-            let view_id2 = view_id.to_owned();
-
-            start_plugin(self_ref, &plugin, plugin_id, &view_id, move |result| {
-                match result {
-                    Ok(plugin_ref) => {
-                        plugin_ref.initialize(&init_info);
-                        let mut inner = me.lock();
-                        let mut running = false;
+        let _ = match self.running_for_view(view_id) {
+            Ok(plugins) if plugins.contains_key(plugin_name) => {
+                Err(Error::Other(format!("{} already running", plugin_name)))
+            }
+            Err(err) => Err(err),
+            Ok(_) => Ok(()),
+        }?;
 
-                        if let Some(ed) = inner.buffers.lock()
-                            .editor_for_view(&view_id2) {
-                                ed.plugin_started(&view_id2, &key.1);
-                                running = true;
-                        }
-                        if running {
-                            inner.running.insert(key, plugin_ref);
+        let plugin_id = self.next_plugin_id();
+        let plugin = self.catalog.iter()
+            .find(|desc| desc.name == plugin_name)
+            .ok_or(Error::Other(format!("no plugin found with name {}", plugin_name)))?;
+
+        let me = self_ref.clone();
+        let view_id2 = view_id.to_owned();
+        let plugin_name = plugin_name.to_owned();
+
+        start_plugin_process(self_ref, &plugin, plugin_id, &view_id, move |result| {
+            let view_id = view_id2;
+            match result {
+                Ok(plugin_ref) => {
+                    plugin_ref.initialize(&init_info);
+                    let mut inner = me.lock();
+                    let mut running = false;
+
+                    if let Some(ed) = inner.buffers.lock()
+                        .editor_for_view(&view_id) {
+                            ed.plugin_started(&view_id, &plugin_name);
+                            running = true;
                         }
```

Comment:
That thing where your text editor keeps wanting to indent close brackets like these incorrectly. This could also be solved by making the if condition a single line, it looks like it fits, it wouldn't even be the longest line in the file.